client/pluginmanager/csimanager/instance.go

```diff
@@ -0,0 +1,209 @@
+package csimanager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/dynamicplugins"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/csi"
+)
+
+const managerFingerprintInterval = 30 * time.Second
+
+type instanceManager struct {
+	info   *dynamicplugins.PluginInfo
+	logger hclog.Logger
+
+	updater UpdateNodeCSIInfoFunc
+
+	shutdownCtx         context.Context
+	shutdownCtxCancelFn context.CancelFunc
+	shutdownCh          chan struct{}
+
+	fingerprintNode       bool
+	fingerprintController bool
+
+	client csi.CSIPlugin
+}
+
+func newInstanceManager(logger hclog.Logger, updater UpdateNodeCSIInfoFunc, p *dynamicplugins.PluginInfo) *instanceManager {
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	return &instanceManager{
+		logger:  logger.Named(p.Name),
+		info:    p,
+		updater: updater,
+
+		fingerprintNode:       p.Type == "csi-node",
+		fingerprintController: p.Type == "csi-controller",
+
+		shutdownCtx:         ctx,
+		shutdownCtxCancelFn: cancelFn,
+		shutdownCh:          make(chan struct{}),
+	}
+}
+
+func (i *instanceManager) run() {
+	c, err := csi.NewClient(i.info.ConnectionInfo.SocketPath)
+	if err != nil {
+		i.logger.Error("failed to setup instance manager client", "error", err)
+		close(i.shutdownCh)
+		return
+	}
+	i.client = c
+
+	go i.runLoop()
+}
+
+func (i *instanceManager) requestCtxWithTimeout(timeout time.Duration) (context.Context, context.CancelFunc) {
+	return context.WithTimeout(i.shutdownCtx, timeout)
+}
+
+func (i *instanceManager) runLoop() {
+	// basicInfo holds a cache of data that should not change within a CSI plugin.
+	// This allows us to minimize the number of requests we make to plugins on each
+	// run of the fingerprinter, and reduces the chances of performing overly
+	// expensive actions repeatedly, and improves stability of data through
+	// transient failures.
+	var basicInfo *structs.CSIInfo
+
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-i.shutdownCtx.Done():
+			if i.client != nil {
+				i.client.Close()
+				i.client = nil
+			}
+			close(i.shutdownCh)
+			return
+		case <-timer.C:
+			ctx, cancelFn := i.requestCtxWithTimeout(30 * time.Second)
```

Comment:
We're passing this context into `buildBasicFingerprint`, which seems to block on talking to the plugin if I understand correctly. Should this timeout match the `managerFingerprintInterval`?

Reply:
oh good spot! - forgot to update this when I was moving stuff to consts